proto/pulumi/provider.proto

```diff
@@ -115,13 +115,23 @@ service ResourceProvider {
     // Call dynamically executes a method in the provider associated with a component resource.
     rpc Call(CallRequest) returns (CallResponse) {}
 
-    // Check validates that the given property bag is valid for a resource of the given type and returns the inputs
-    // that should be passed to successive calls to Diff, Create, or Update for this resource. As a rule, the provider
-    // inputs returned by a call to Check should preserve the original representation of the properties as present in
-    // the program inputs. Though this rule is not required for correctness, violations thereof can negatively impact
-    // the end-user experience, as the provider inputs are using for detecting and rendering diffs.
+    // `Check` validates a set of input properties against a given resource type. A `Check` call returns either a set of
+    // checked, known-valid inputs that may subsequently be passed to [](pulumirpc.ResourceProvider.Diff),
+    // [](pulumirpc.ResourceProvider.Create), or [](pulumirpc.ResourceProvider.Update); or a set of errors explaining
+    // why the inputs are invalid. In the case that a set of inputs are successfully validated and returned, `Check`
+    // *may also populate default values* for resource inputs, returning them so that they may be passed to a subsequent
+    // call and persisted in the Pulumi state. In the case that `Check` fails and returns a set of errors, it is
+    // expected that the caller (typically the Pulumi engine) will fail resource registration.
+    //
+    // As a rule, the provider inputs returned by a call to `Check` should preserve the original representation of the
+    // properties as present in the program inputs. Though this rule is not required for correctness, violations thereof
+    // can negatively impact the end-user experience, as the provider inputs are used for detecting and rendering
+    // diffs.
     rpc Check(CheckRequest) returns (CheckResponse) {}
-    // Diff checks what impacts a hypothetical update will have on the resource's properties.
+
+    // `Diff` compares an existing ("old") set of resource properties with a new set of properties and computes the
+    // difference (if any) between them. `Diff` should only be called with values that have at some point been validated
+    // by a [](pulumirpc.ResourceProvider.Check) call.
```

Comment:
Diff receives old state, old checked input, new checked input. I wonder if an implementer would benefit from the suggestion that they use a three-way merge strategy when computing a diff. Akin to this: https://github.com/kubernetes/apimachinery/blob/a4471387da34d0c88594d7620508e02169c806a7/pkg/util/strategicpatch/patch.go#L2094-L2105

Without guidance, they might resort to computing the diff between old/new input, which would fail to detect to drift. Or they might do new input to old state, which would fail to detect the "unsetting" of a input.